### ISSUE TITLE
[Feature] Add Support for Choppers/Sugcane back in

### ISSUE THREAD

**Issue by pops64** [Courseplay/Courseplay_FS22#2701]
CP as always supported choppers since at least FS15(When I started playing). I would love to see them brought back in with support. Most of the code to support this is already there just needs a few tweaks to handle some of the particularities of  Forage Harvester Unloading.

Looked through the code base and from experience in 17. The missing items to add this functionality back in are as follows.

- [x] Combine Driver: Needs to wait until tipper is in range unless turning(Code not present) GDNs [website](https://gdn.giants-software.com/documentation_scripting_fs22.php?version=script&category=1&class=4) on how they do this
- [x] Combine Unloading Driver Needs to handle end of row turn around. Moving out of the way and looping around with enough space to realign with Chopper. The tipper needs be checked if reversible this has and can cause issues with turn maneuvers. It appears most of the code needed to achieve this is already present in CP just need to combine the parts together 
- [x] Combine Unloading Driver needs to handle headlands or disable it complete. Any sharp curve in the headland maneuver caused issues in FS17. The only time CP FS17 was able to handle headlands with choppers where when the course gen was set to Sharp forcing the driver to handle it like a turn. Otherwise it wouldn't be able to stay in range of the chopper or it would collide depending on the curve
- [x] Shifting the Combine Unloading Driver call to chopper when current tipper is almost full instead of the combine as currently coded(Not sure if this is coded when multiple tippers are unloading a combine)
- [x] Choppers can unload both left and right side. An option to pick which side or just always pick the side without fruit(No option always non fruit side 95% of the time)

I really hope this can be added back in. As babysitting a chopper is a PITA and time can be better spent else where. I see autodrive can handle? Never really used AD it was always clunky and I always enjoyed CP being the one stop shop of FS helper automation

**Comment by Tensuko** [comment]
Nothing will be imported from FS17.
If this will be put back into FS22, will import it from FS19 but it still needs more adjustments and needs to be put into a chopperUnloaderDriver to match our new file organisation and not end up as chaotic as in FS19 and before.

Chopper is waiting untill can unload, not sure if there is a big difference, might can adjust this.

The rest will come from FS19 + adjustments that are needed.

However it was never planed so can't rly say if we put it back in.
Maybe next year? Who knows....

There is other stuff that needs to be done first.

**Comment by pops64** [comment]
Sorry if this is out of place. For those interested in following this progress. I am including support for this on my fork. It is very much an alpha so I won't link it but for those who know how to find it I trust that you understand that it may cause game breaking bugs. 
So far I have gotten the chopper to drive just fine. And 1 tipper to follow it. Turns will occur but they are janky. Things I haven't figured out is unloading. Currently unloading to field works just fine maybe there is a way to use this same method to silo bunker

**Comment by KCHARRO** [comment]
In FS17 I remember, in Lapacho, I put 3 small tractors to unload a Sugarcane Combine and it worked well, only rarely did an unforeseen event occur.  Sometimes i had to go out of the house and leave them alone, when i returned the job was done without any problems.

 I miss that......

**Comment by pops64** [comment]
That is surprisings Sugarcane choppers where always a little on the flaky side. Could accelerate faster than the tractor causing issues. Not to mention their small working width caused issues

**Comment by KCHARRO** [comment]
I always used the smallest Hürlimann tractor (the H488)with two trailers, the smallest ones, and also this mod "Tatra Phoenix Agro Pack" also with 3 trucks and a trailer "of the pack itself" each one, and it really gave me few(or almost no problem since the Tatra is a small truck with a tight turning circle... Now that I think about it, it may be the problem, the minimum turning radius of the vehicle that is used as an unloader).

At that time I had the PC next to the bed, I put them to work and slept a little in the afternoons, when I woke up the work was already done or almost finished(something I missed in FS19, now in FS22), I still have everything that, there are 171 mods that I have for FS17.

**Comment by Tensuko** [comment]
Nothing from FS17 will be taken over.

**Comment by pops64** [comment]
Yeah I wasn't around for FS19 not sure if there was chopper support. My goal is to code in the chopper support following CP FS22 standard. A little teaser
![20230816161317_1](https://github.com/Courseplay/Courseplay_FS22/assets/10678220/8c139c37-6d54-4090-8f82-086e5df56344)
![20230816171831_1](https://github.com/Courseplay/Courseplay_FS22/assets/10678220/83d3c8d9-f755-4aae-b78e-43d17db7cedf)

There is still a bunch more testing to be done. Currently it will only unload 1 tipper at a time. I plan on adding code so it will call a 2nd driver when the first is close to full. I am thinking I can use the same trailer unload setting for combine unloaders to determine when this should occur. Then a bunch of testing of combines/choppers/sugarcane to be done. There was some under the hood stuff that may impact combines nothing functionally has changed just want to make sure I didn't break anything(What chopper code there was. Was included in the combine code). I may add code so CP can unload into a bunker silo but this may or may not happen as CP currently really doesn't support this functionality any more so I would more than likely have to make it from starch. Plus I have heard AD can handle silo bunkers.

Also I may change the turn around code. It is 50k time more reliable that FS17 but it is relying on pathfinder so it is an off the field turn.

So far Turns work, Fruit Side Switching works, and unloading on field

**Comment by Tensuko** [comment]
Just use the pathfinder for it. Off field is only a problem for simple up down course.
Also check follow chopper through turn state of FS19 for longer turns of chopper.

**Comment by pops64** [comment]
It wouldn't be course recording. I haven't even looked at how that has changed.(I haven't really figured it out in game other than custom fields). My plan would be to use the giants go to function(CP already supports this) Drive to the bunker use pathfinder to unload into the bunker. The silo loader/lever appears to do this already I would just have to see if it can handle a trailer. I have high hopes as this is how unload on field is done as well. And then use giants go to function to go back to field.(CP also already supports this). I only have two problems if the bunker is in a constrained area I am not sure how well pathfinder would handle it. I also am not sure where to put this feature on the create job page as it is already quite full for combine unload.

Either way that is on my bottom of my todo list. I probably will make a pull request once I get 2 tipper drivers supported, clean the code up and do testing(I probably will need help with this, too many different edge cases to think of).

Off the field turns work fine unless you have the field right up against the fence line. This is the one edge case that will break chopper support. I am not to sure yet on how to handle this. This edge case was always a problem in 17. If I have headlands I can ignore fruit side and place the chopper on the inside. The problem with this are the corners of the field. If they are round the chopper probably will collide(Maybe not any more seeing how the tipper has the combine course). If they are sharp the turn maneuver will be a problem. Tho there is code to get out of the way so this may not be an issue

**Comment by rohne83** [comment]
Rather not reengage CP at all anywhere outside the field, also not in the bunker. If you want to unload into bunker, then AD has to do this. Maybe Giants helper will support bunkers as well anytime in the future.
Unloading into bunker by CP pathfinder might create loads of unforeseen issues.

**Comment by pops64** [comment]
It will have to be the AD. I have tested it and the hand off to AD works fine. Giants helper only delivers to sell points it doesn't recognize bunker silos. I am currently working on getting a tipper to chase a currently unloading chopper so it's close by when the current unloader is full. Once that is done all that will be left is testing. If your interested in looking at what this might be there is a draft PR with a list of things that should be tested.

**Comment by pops64** [comment]
A little update. Mostly feature complete just need to do testing and approval from this CP team to merge this. There are some minor tweaks and adjustments needed to be made(Handling chopper turn around with 2 tippers waiting and the chopper telling the 2nd tipper to arrive on the next row)
![20230822111520_1](https://github.com/Courseplay/Courseplay_FS22/assets/10678220/b3bdc724-157f-443e-b52c-f83bcef29274)
![20230822111538_1](https://github.com/Courseplay/Courseplay_FS22/assets/10678220/26e68f8d-83a6-4370-a7c6-0528955c6cf9)

**Comment by pops64** [comment]
Due to differing conflicting interests this will no longer be added back to CP unless it is done by the CP owner. I am no longer interested in this support so this will be closed. I will leave my fork this active for those that want to merge my code base into yours. It was mostly feature complete needed some minor tweaks and some logic for better handling of avoiding fruit when departing on a full load is missing but it will unload the chopper just fine. It can't handle sugarcane unfortunately.

**Comment by pops64** [comment]
![20230825202350_1](https://github.com/Courseplay/Courseplay_FS22/assets/10678220/5c381cb5-740c-4a74-81cb-7ff8870640e5)
Currently kinda of breaks on headlands. Due to the weird pipe angle and height of trailer and just the closeness of everything. Up/Down Rows work just great!

**Comment by Misamangame** [comment]
Looks great, can those trailers unload automatically either with CP/AD/Ghelper?

**Comment by pops64** [comment]
Unfortunately the sugar cane trailer unload code never made it to FS22 CP. It's something I would have to program in but due the fact that in FS19 it used user supplied settings for the unload position which also isn't in current CP. I would have to go max rotation on both for it to unload which may or may not be what the current CP team wants but for right now I am only trying to focus on Choppers in general.

**Comment by Tensuko** [comment]
Got a plan for the sugar cane trailers, but we will handle them later.

**Comment by pops64** [comment]
Alright I am looking for testers as I am very close to this being finalized for approval by the CP team. Let me know of any bugs on the [PR](https://github.com/Courseplay/Courseplay_FS22/pull/2712) for this. You can check out it there or on my [fork](https://github.com/pops64/Courseplay_FS22/tree/ChopperSupport). Thanks for any feed back

**Comment by regs1980** [comment]
Bugs I have found part 1

pathfinder cannot find the correct point

https://drive.google.com/file/d/1s96k2VlKQ9p0TweULRg3_d7ELwljJ95v/view?usp=drive_link
https://drive.google.com/file/d/1z_IziTRrB1OsxxwCWVWiOM0A_2X-mDOW/view?usp=drive_link

Why does the driver turn the loop before?
https://drive.google.com/file/d/1mVATuMzm2iR_CmeCOBWw4-wfG0VNNjbM/view?usp=drive_link

Active CP driver impedes the chopper driving
https://drive.google.com/file/d/1RheDuCY7L2s-iCnhCo-ptqMWENeBCFG3/view?usp=drive_link

CP Driver stays behind the Choppers at the 2nd headland
https://drive.google.com/file/d/1LJ0A6db4ws2aa-n1g3picLlGHkqrKtaF/view?usp=drive_link
CP driver can drive alongside the chopper when he is in the 2nd headland

Chopper calls too many drivers 
https://drive.google.com/file/d/17y3zbthcNuJ6nZXD1ECLOYDOI6yubbyE/view?usp=sharing
No second driver has to be called as long as the filling volume has not yet been reached.

The distance between the chopper and the fruit is too large, the driver can't find a way through. 
https://drive.google.com/file/d/1ywi3XSSuAfk98UxTBGMo9lHiOwClWlUR/view?usp=drive_link
https://drive.google.com/file/d/1ytFi0M3JQo0N7oD2v6OEzXKycR-2Yz6j/view?usp=drive_link
https://drive.google.com/file/d/1F7IRENe-5M6SPmJ-ZbbDJ0Jgi68PbSjp/view?usp=drive_link

The Pathfinder makes too big loops and does not recognize obstacles correctly.
https://drive.google.com/file/d/1fnzoUmnUIXFnzzCbxKQaLgs1FmrdJT3W/view?usp=drive_link

**Comment by rohne83** [comment]
I've tried out your PR. As long as it's working it is definitely a nice addition and much more fluent than using AD unloaders. But I also observed few issues:
1. missing l10n entries. Makes the Hud nearly unusable.
![fsScreen_2023_09_02_10_05_30](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/1320742a-5379-4981-8d7a-2e4a126a5389)

2. It happened sometimes (especially on headland turns) that two unloaders (the one following the chopper and the next one) block each other, and then everything was blocked and it was quite hard to get the chopper running again at all. Just disengage+reengage didn't really help. Had to stop all other unloaders, drive first one in to better postion and reengage, and then stop+reengage the chopper again as well. Quite hard to figure out what caused the issue
![fsScreen_2023_09_02_10_15_07](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/783fe3e8-52e5-4317-8c5a-c34f8370024f)

3. sharp headland turns by unloaders were mostly performed outside the field instead of staying within field boundaries
![fsScreen_2023_09_02_10_01_45](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/e4707a21-9080-4840-94b4-0749a56026e5)

4. don't really understand when the second unloader is being called. On headland (at least during first pass) the second unloader was never called until the first one was completely full. On later rows it seemed to be related to the fill level set for previous unloader. But if you set this to a low value i.e. 50% so that second unloader will be there quite early (might be necessary especially on large fields with long driving distances), this also resulted in the first unloader stopping half full as soon as second unloader is near. Behaviour in FS19 (and still now with AD), that current unloader keeps following chopper until the next turn or completely full (and second unloader just following first unloader during that time) makes more sense in my eyes.

5. hand over to AD is very strange. Sometimes on first headland it was handed over directly, later on field rows it first performed a u-turn away from the chopper (quite nice if the rest of the handover worked better), sometimes the unloader went back to the startpoint first (directly or after that u-turn) which could create issues with following unloaders or AI traffic and also took very long sometimes for pathfinder, sometimes (especially during second+third headland pass) the unloader even didn't hand over to AD at all but tried to drive around the whole map using CP pathfinder. All those behaviours could be observed with any unloader although all had same AD+CP settings and no setting performed in AI menue (except for choosing the field with the target cross)
![fsScreen_2023_09_02_10_26_19](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/02f55c79-40fa-4b20-aa52-a3fdc9efcffa)
![fsScreen_2023_09_02_10_28_26](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/f2f500d5-7303-4f2a-b69c-0f52df202c84)
![fsScreen_2023_09_02_11_04_48](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/2ff953e0-2afc-47c1-8087-ba0dbe99fb1b)
![fsScreen_2023_09_02_10_39_42](https://github.com/Courseplay/Courseplay_FS22/assets/71030138/7cab18aa-bdda-47be-958d-6ecd2d9772e5)

**Comment by pops64** [comment]
Thanks for the feed back. 
@regs1980 
- Pathfinder can't find correct point. I haven't yet tested sharp headland turns. It will be something I have to figure out how to handle. And tell the chopper not to make the pocket for the turn(You are losing fruit cause of this)
- The loop. With out debuging on I couldn't say for sure. The code relies on pathfinder to get there.
- Was the 3rd chopper a CP driver in chopper unload? I will have to disable the code that its looking for it's current driver. Unless I can lock it on one target. Doubtful tho
- When it changes headlands. I never thought to code an update for where it wants to drive when headlands row change. That's why in your next vid the 2nd and 3rd driver go beside it
- I would need debugging on with the log all vehicles chopper, and all 3 unloaders. I will see if I can reproduce it
- Fruit point to large. See first point. 
- That actually isn't pathfinders fault. I manually create a course to get away from the chopper. It's suppose to check to see if that turn maneuver would put it outside of the field boundary and if not drive straight unless it is on a headland. I will have to update to follow the choppers course x amount of distance when on a headland. Sometimes pathfinder takes to long following the choppers course isn't always the best option for the next unloader 

@rohne83 
1. It's the way CP works with translations now. They have a master file that when it gets pushed onto the main branch it is copied over to all langue files. Easier on CP team not having to manually add 1 line to the 12? or so files. If you set your language to English it should be fine or add the line from the English xml to your language xml file
2. Yeah I didn't program sharp turn logic at all for the unload drivers. Sorry :(
3. See above
4. It uses the same logic as the combine. When it figures out how quickly it will reach the fill level setting for the unloader(It goes by the current unloaders fill level setting) And finds a driver that will reach it in that time. There's a lot of headaches involved with trying to have two unloaders follow each other. I will take it look at it again and see if I can get something working.
5. There are 4 currently programed maneuvers. 
- The first one is for when it is behind the chopper it will either back up straight 
- Or when it is a land row(fruit both sides of the lane) backup and make a 180(Turn circle is too wide currently will stop right before it hits the chopper). And end up back in the worked row.(Pathfinder takes way to long when point in the wrong direction)
- It will drive straight to the end of the row if there isn't enough room to make a wide 180
-  A wide 180 
I did notice in the first two of your pictures it drove straight off the field I will look into getting this fixed.(I switched from a fixed distance to going with the length of the row remaining to drive straight). After that it is the same code used by combines. Try again with the same field using a combine and combine driver and see if your can reproduce your results. If you can create a bug report about. 

Thanks again for taking your time and testing this. Defiantly some cases I never thought to look or completely forgot about. I will hopefully have time to get theses edge cases fixed and updated pushed to my fork

**Comment by zinni97** [comment]
Thanks for your work I have been waiting and hoping for that mode to come back! I have been testing it quite a bit as well and it works much smoother overall than AD. 
I have also had the issue #5 that Rohne83 with the handover being weird and the unloader wanting to go in a straight line to nowhere. 

I also had occasional crashes with the log error _:686: attempt to index field 'combine' (a nil value)
2023-09-03 22:42 Error: Running LUA method 'update'._
**Important note:** Not sure this is caused by the fork though! I have been running it on my regular savegame which isn't the most stable.

**Comment by Tensuko** [comment]
Dont get your hopes up too high.
I have minimum requirments for this PR and right now it doesnt look good.
The first headland turn is still not working correct and "chamge the chopper turn" is not the correct answer to this.
Transition from chase on 1st headland to 2nd headland drive beside also not working.
2nd unloader call is bugged at turns and stick down all other waiting unloader is also not the way to go.

If this doesnt want to be changed the correct way, well it will never make its way into CP like this.

**Comment by pops64** [comment]
@Tensuko The change from chase to drive beside might not happen. I will look through 19 and see if/how they implemented but it would be a headache as there is safety check code that if the unloader gets out of line it tries to get back in. Usually takes 5-10 secs cause of the combine code works and require a major investment in order to speed this up. I am going to disable the pocket mode for the chopper turn. But not the chopper turn it will turn like any fieldworker. There is no way to avoid this. Like I have said earlier I am not going to enable cheats just to make it do something you want.(I did test this take the longest tractor, articulated is the best, and try to chase a chopper it physical isn't possible unless you cheat)

> 2nd unloader call is bugged at turns and stick down all other waiting unloader is also not the way to go.

This makes no sense at all.

Also I am not going to make a duplicate FS19 chopper for various reasons.  The biggest is this is FS22 and FS22 CP there have been a lot of under the hood changes that make this impossible(Getting chase made working correctly was a huge task as there was a lot of upstream debugging to do)

What I will do is make something that I will play myself personally as that is my standard. If there is something I don't like I will either figure out how to program a solution or make it as easy on the user to workaround. As I am very lazy when it comes to FS and would have up to 8 workers on one field in FS17

@zinni97 Do have the rest of that debug line there should be more text before the :686 that is a reference to a line of code it should have the file that is associated to it. I do know occasionally there is a non game breaking nil error but for me it is printing a call stack and this particular stack isn't helpful to its location

**Comment by Tensuko** [comment]
Then I don't see any reason why this should be continued.
If you not accepting our requiered changes and make things even worse with the chopper, this will never find its way into CP.

As this seems to be your final answer to this, I will end this here and now.
Sadly...
